### PR TITLE
yegor-sereda-dao3-hw

### PR DESCRIPTION
**Pull request by entertainment-veks**
added Car, Driver and Manufacturer models
added ManufacturerDao + impl
added ManufacturerService + impl
added Injector
added controller (Application.java)

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -8,5 +8,60 @@
...
+                    <configuration>
+                        <source>${jdk.version}</source>
+                        <target>${jdk.version}</target>
+                        <encoding>${project.build.sourceEncoding}</encoding>
+                    </configuration>
+                </plugin>
+            </plugins>
+        </pluginManagement>
+    </build>
 </project>
```
Comment (Fedich09) [inline pom.xml:67]:
```suggestion
</project>

```
add empty line please

File: src/main/java/dao3/dao/impl/ManufacturerDaoImpl.java [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+
+    @Override
+    public List<Manufacturer> getAll() {
+        return Storage.manufacturers;
+    }
+
+    @Override
+    public Manufacturer update(Manufacturer manufacturer) {
+        for (int i = 0; i < Storage.manufacturers.size(); i++) {
+            if (Storage.manufacturers.get(i).getId() == manufacturer.getId()) {
```
Comment (Fedich09) [inline src/main/java/dao3/dao/impl/ManufacturerDaoImpl.java:32]:
```suggestion
            if (Storage.manufacturers.get(i).getId() == manufacturer.getId()) {
```
you should compare objects with equals

File: src/main/java/dao3/dao/impl/ManufacturerDaoImpl.java [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+
+    @Override
+    public boolean delete(Long id) {
+        Optional<Manufacturer> current =
+                Storage.manufacturers.stream()
+                .filter(m -> m.getId() == id)
+                .findFirst();
+
+        return current.map(m -> Storage.manufacturers.remove(m)).orElse(false);
+
```
Comment (Fedich09) [inline src/main/java/dao3/dao/impl/ManufacturerDaoImpl.java:48]:
```suggestion
        return Storage.manufacturers.remove(id);
```

File: src/main/java/dao3/model/Manufacturer.java [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+        this.country = country;
+    }
+
+    @Override
+    public String toString() {
+        return "Manufacturer{"
+                + "id=" + id + ", name='" + name + '\''
+                + ", country='" + country + '\'' + '}';
+    }
+}
```
Comment (Fedich09) [inline src/main/java/dao3/model/Manufacturer.java:43]:
```suggestion
}
```
and you should add equals and hashCode methods

**Review by Fedich09 — CHANGES_REQUESTED** [review]

**Review by Fedich09 — APPROVED** [review]

File: pom.xml [hunk 5]
```diff
@@ -8,5 +8,60 @@
...
+        <maven.checkstyle.plugin.configLocation>
+            https://raw.githubusercontent.com/mate-academy/style-guides/master/java/checkstyle.xml
+        </maven.checkstyle.plugin.configLocation>
+    </properties>
 
-</project>
\ No newline at end of file
+    <dependencies>
+        <dependency>
+            <groupId>junit</groupId>
```
Comment (Sofasmile) [inline pom.xml:23]: do we need this dependency here?

File: src/main/java/dao3/dao/impl/ManufacturerDaoImpl.java [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+public class ManufacturerDaoImpl implements ManufacturerDao {
+    @Override
+    public Manufacturer create(Manufacturer manufacturer) {
+        return Storage.addManufacturerWithId(manufacturer);
+    }
+
+    @Override
+    public Optional<Manufacturer> get(Long id) {
+        return Storage.manufacturers.stream()
+                .filter(m -> m.getId() == id)
```
Comment (Sofasmile) [inline src/main/java/dao3/dao/impl/ManufacturerDaoImpl.java:20]: what is the difference berween == and equals?

File: src/main/java/dao3/dao/impl/ManufacturerDaoImpl.java [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+
+    @Override
+    public List<Manufacturer> getAll() {
+        return Storage.manufacturers;
+    }
+
+    @Override
+    public Manufacturer update(Manufacturer manufacturer) {
+        for (int i = 0; i < Storage.manufacturers.size(); i++) {
+            if (manufacturer.getId() == Storage.manufacturers.get(i).getId()) {
```
Comment (Sofasmile) [inline src/main/java/dao3/dao/impl/ManufacturerDaoImpl.java:32]: the same

File: src/main/java/dao3/dao/impl/ManufacturerDaoImpl.java [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+                Storage.manufacturers.set(i, manufacturer);
+                return manufacturer;
+            }
+        }
+        return null;
```
Comment (Sofasmile) [inline src/main/java/dao3/dao/impl/ManufacturerDaoImpl.java:37]: return null is a bad practice, better throw exception

File: src/main/java/dao3/dao/impl/ManufacturerDaoImpl.java [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+    }
+
+    @Override
+    public boolean delete(Long id) {
+        return Storage.manufacturers.removeIf(e -> e.getId() == id);
```
Comment (Sofasmile) [inline src/main/java/dao3/dao/impl/ManufacturerDaoImpl.java:42]: the same

File: src/main/java/dao3/db/Storage.java [hunk 10]
```diff
@@ -0,0 +1,38 @@
+package dao3.db;
+
+import dao3.model.Car;
+import dao3.model.Driver;
+import dao3.model.Manufacturer;
+import java.util.ArrayList;
+import java.util.List;
+
+public class Storage {
+    public static List<Car> cars = new ArrayList<>();
```
Comment (Sofasmile) [inline src/main/java/dao3/db/Storage.java:10]: let's add final keyword to all lists

File: src/main/java/dao3/model/Car.java [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+import java.util.List;
+import java.util.Objects;
+
+public class Car {
+    private int id;
+    private String model;
+    private Manufacturer manufacturer;
+    private List<Driver> drivers;
+
+    public Car(int id, String model, Manufacturer manufacturer, List<Driver> drivers) {
```
Comment (Sofasmile) [inline src/main/java/dao3/model/Car.java:12]: should we pass id through constructor? 
the same about drivers?
please, read checklist - https://mate-academy.github.io/jv-program-fulltime/02_jdbc/checklist/03_dao_service_checklist.html

File: src/main/java/dao3/model/Car.java [hunk 12]
```diff
@@ -0,0 +1,74 @@
+package dao3.model;
+
+import java.util.List;
+import java.util.Objects;
+
+public class Car {
+    private int id;
```
Comment (Sofasmile) [inline src/main/java/dao3/model/Car.java:7]: id should be Long and wrapper

File: src/main/java/dao3/model/Driver.java [hunk 13]
```diff
@@ -0,0 +1,61 @@
+package dao3.model;
+
+import java.util.Objects;
+
+public class Driver {
+    private long id;
```
Comment (Sofasmile) [inline src/main/java/dao3/model/Driver.java:6]: and here

File: src/main/java/dao3/model/Driver.java [hunk 14]
```diff
@@ -0,0 +1,61 @@
...
+    private String name;
+    private String licenceNumber;
+
+    public Driver(long id, String name, String licenceNumber) {
```
Comment (Sofasmile) [inline src/main/java/dao3/model/Driver.java:10]: the same

File: src/main/java/dao3/model/Manufacturer.java [hunk 15]
```diff
@@ -0,0 +1,60 @@
+package dao3.model;
+
+import java.util.Objects;
+
+public class Manufacturer {
+    private long id;
```
Comment (Sofasmile) [inline src/main/java/dao3/model/Manufacturer.java:6]: and here

File: src/main/java/dao3/service/impl/ManufactererServiceImpl.java [hunk 16]
```diff
@@ -0,0 +1,39 @@
...
+import dao3.lib.Inject;
+import dao3.lib.Service;
+import dao3.model.Manufacturer;
+import dao3.service.ManufacturerService;
+import java.util.List;
+
+@Service
+public class ManufactererServiceImpl implements ManufacturerService {
+    @Inject
+    ManufacturerDao manufacturerDao;
```
Comment (Sofasmile) [inline src/main/java/dao3/service/impl/ManufactererServiceImpl.java:13]: how about access modifier?

File: src/main/java/dao3/service/impl/ManufactererServiceImpl.java [hunk 17]
```diff
@@ -0,0 +1,39 @@
...
+
+    @Override
+    public Manufacturer create(Manufacturer manufacturer) {
+        return manufacturerDao.create(manufacturer);
+    }
+
+    @Override
+    public Manufacturer get(Long id) {
+        return manufacturerDao.get(id).orElse(null);
```
Comment (Sofasmile) [inline src/main/java/dao3/service/impl/ManufactererServiceImpl.java:22]:
```suggestion
        return manufacturerDao.get(id).get();
```

**Review by Sofasmile — CHANGES_REQUESTED** [review]

File: src/main/java/dao3/dao/impl/ManufacturerDaoImpl.java [hunk 18]
```diff
@@ -29,16 +29,16 @@ public Manufacturer create(Manufacturer manufacturer) {
...
     public Manufacturer update(Manufacturer manufacturer) {
         for (int i = 0; i < Storage.manufacturers.size(); i++) {
-            if (manufacturer.getId() == Storage.manufacturers.get(i).getId()) {
+            if (manufacturer.getId().equals(Storage.manufacturers.get(i).getId())) {
                 Storage.manufacturers.set(i, manufacturer);
                 return manufacturer;
             }
         }
-        return null;
+        throw new RuntimeException("Can't update, element doesn't exist");
```
Comment (Sofasmile) [inline src/main/java/dao3/dao/impl/ManufacturerDaoImpl.java:37]:
```suggestion
        throw new RuntimeException("Can't update, element " + manufacturer + " doesn't exist");
```

File: src/main/java/dao3/model/Car.java [hunk 19]
```diff
@@ -1,26 +1,27 @@
...
 public class Car {
-    private int id;
+    private Long id;
     private String model;
     private Manufacturer manufacturer;
     private List<Driver> drivers;
 
-    public Car(int id, String model, Manufacturer manufacturer, List<Driver> drivers) {
+    public Car(Long id, String model, Manufacturer manufacturer) {
         this.id = id;
```
Comment (Sofasmile) [inline src/main/java/dao3/model/Car.java]:
```suggestion
    public Car(String model, Manufacturer manufacturer) {
```

File: src/main/java/dao3/model/Driver.java [hunk 20]
```diff
@@ -0,0 +1,61 @@
...
+import java.util.Objects;
+
+public class Driver {
+    private Long id;
+    private String name;
+    private String licenceNumber;
+
+    public Driver(long id, String name, String licenceNumber) {
+        this.id = id;
+        this.name = name;
```
Comment (Sofasmile) [inline src/main/java/dao3/model/Driver.java:12]:
```suggestion
    public Driver(String name, String licenceNumber) {
        this.name = name;
```

**Review by Sofasmile — CHANGES_REQUESTED** [review]

**Review by Sofasmile — APPROVED** [review]

**Pull request merged**